src/RamCondition.h

```diff
@@ -31,138 +31,208 @@
     std::unique_ptr<RamValue> lhs;
-    /** right-hand side */
+
+    /** Right-hand side of constraint */
     std::unique_ptr<RamValue> rhs;
 
 public:
     RamBinaryRelation(BinaryConstraintOp op, std::unique_ptr<RamValue> l, std::unique_ptr<RamValue> r)
             : RamCondition(RN_BinaryRelation), op(op), lhs(std::move(l)), rhs(std::move(r)) {}
 
-    ~RamBinaryRelation() override = default;
-
+    /** Print */
     void print(std::ostream& os) const override {
         lhs->print(os);
         os << " " << toBinaryConstraintSymbol(op) << " ";
         rhs->print(os);
     }
+
+    /** Get level */
     size_t getLevel() override {
         return std::max(lhs->getLevel(), rhs->getLevel());
     }
-    /** get left-hand side */
+
+    /** Get left-hand side */
     RamValue* getLHS() const {
         return lhs.get();
     }
-    /** get right-hand side */
+    /** Get right-hand side */
     RamValue* getRHS() const {
         return rhs.get();
     }
 
+    /** Take left-hand side */
     std::unique_ptr<RamValue> takeLHS() {
         return std::move(lhs);
     }
 
+    /** Take right-hand side */
     std::unique_ptr<RamValue> takeRHS() {
         return std::move(rhs);
     }
 
-    /** set LHS */
+    /** Set left-hand side */
     void setLHS(std::unique_ptr<RamValue> l) {
         lhs.swap(l);
     }
-    /** set RHS */
+    /** Set right-hand side */
     void setRHS(std::unique_ptr<RamValue> r) {
         rhs.swap(r);
     }
 
+    /** Get operator symbol */
     BinaryConstraintOp getOperator() const {
         return op;
     }
 
-    /** Obtains a list of child nodes */
+    /** Obtain list of child nodes */
     std::vector<const RamNode*> getChildNodes() const override {
         return {lhs.get(), rhs.get()};
     }
+
+    /** Create clone */
+    RamBinaryRelation* clone() const override {
+        RamBinaryRelation* res = new RamBinaryRelation(
+                op, std::unique_ptr<RamValue>(lhs->clone()), std::unique_ptr<RamValue>(rhs->clone()));
+        return res;
+    }
+
+    /** Apply */
+    void apply(const RamNodeMapper& map) override {
+        lhs = map(std::move(lhs));
+        rhs = map(std::move(rhs));
+    }
+
+protected:
+    /** Check equality */
+    bool equal(const RamNode& node) const override {
+        assert(dynamic_cast<const RamBinaryRelation*>(&node));
+        const RamBinaryRelation& other = static_cast<const RamBinaryRelation&>(node);
+        return getOperator() == other.getOperator() && getLHS() == other.getLHS() &&
+               getRHS() == other.getRHS();
+    }
 };
 
-/** check whether a tuple (pattern) does not exist in a relation */
+/** Not existence check for a relation */
 class RamNotExists : public RamCondition {
-    /** the relation to be accessed */
+protected:
+    /* Relation */
     RamRelation relation;
 
-    /** the restricted fields -- null if undefined */
+    /** Pattern -- nullptr if undefined */
+    // TODO: rename to argument
```

Comment:
TODOs should be in the form` // TODO (#issueid): text` or `// TODO (userid): text`

Reply:
Done.